README.md
# ioBroker.steam-friends

<img src="admin/steam-friends.png" height="100">

[![NPM version](https://img.shields.io/npm/v/iobroker.steam-friends.svg)](https://www.npmjs.com/package/iobroker.steam-friends)
[![Downloads](https://img.shields.io/npm/dm/iobroker.steam-friends.svg)](https://www.npmjs.com/package/iobroker.steam-friends)
![Number of Installations (latest)](https://iobroker.live/badges/steam-friends-installed.svg)
![Number of Installations (stable)](https://iobroker.live/badges/steam-friends-stable.svg)
[![Dependency Status](https://img.shields.io/david/xHeitkamp/iobroker.steam-friends.svg)](https://david-dm.org/xHeitkamp/iobroker.steam-friends)
![Vulnerabilities](https://snyk.io/test/github/hk-studio/ioBroker.steam-friends/badge.svg)
![Test and Release](https://github.com/xHeitkamp/ioBroker.steam-friends/workflows/Test%20and%20Release/badge.svg)

[![NPM](https://nodei.co/npm/iobroker.steam-friends.png?downloads=true)](https://nodei.co/npm/iobroker.steam-friends/)

## steam-friends adapter for ioBroker

With this ioBroker Adapter you can see the status of your steam friends

## Configuration

To use this adapter, you need to follow 2 steps:

1. Insert your Steam ID and API key in the corresponding fields in the adapter settings. This data will be used to retrieve the data from Steam.
2. And your Steam friends list must be set to the "Public" setting, otherwise the data cannot be retrieved.
3. (Optional) You can increase the refresh interval of the adapter, if you dont want to DDOS the Steam servers ;) But 10 seconds is perfectly fine!

### Get your Steam ID

The easiest way to get your Steam ID is to use the page: [SteamIDFinder](https://www.steamidfinder.com/).
With this page you can find out the Steam ID you need "steamID64 (Dec)" if you enter your Steam Username. Example: 76561197960287930

### Get your Steam API key

To get your Steam API key you need to already spend more than $5 in the store. If you have done this you can change to the [Steam API key page](https://steamcommunity.com/dev/apikey). On this page you need to confirm the [Steam API Terms of Use](https://steamcommunity.com/dev/apiterms) and enter your Domain Name. Now click "Register" and you should see your key.
Example: M947XROX89UBQ6AOJTOPYJUKAXX6QBTI

### Set your friends list to public

To make your friends list public on Steam, go to the [Profile settings page](https://steamcommunity.com/my/edit/settings) and change the "Friends List" to "Public".

## Changelog

Please read the [CHANGELOG.md](CHANGELOG.md) file for further information

## License

MIT License

Copyright (c) 2023 Robin Behlke (xHeitkamp) <iobroker@hk-studio.de>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


CHANGELOG.md
# Changelog
<!--
    Placeholder for the next version (at the beginning of the line):
    ## **WORK IN PROGRESS**
-->
## 1.0.0 (2023-04-10)

* (xHeitkamp) initial release


lib/helper.test.js
const assert = require('assert');
const helper = require('./helper');

describe('Tests for helper functions', () => {
    //timeConverter()
    describe('#timeConverter', () => {
        it('should return a formatted date string for a valid timestamp', () => {
            const timestamp = 1649605947; // 2022-04-09T17:12:27.000Z
            const expected = '4/10/2022, 5:52:27 PM';
            assert.notStrictEqual(helper.timeConverter(timestamp), expected);
        });

        it('should return an empty string for a timestamp <= 0', () => {
            const timestamp = 0;
            const expected = '';
            assert.strictEqual(helper.timeConverter(timestamp), expected);
        });

        it('should return an empty string for a non-numeric timestamp', () => {
            const timestamp = 'not a timestamp';
            const expected = '';
            // @ts-ignore
            assert.strictEqual(helper.timeConverter(timestamp), expected);
        });
    });

    //orderObject()
    describe('#orderObject', () => {
        it('should return an object with keys in alphabetical order', () => {
            const unordered = { z: 1, y: 2, x: 3 };
            const expected = { x: 3, y: 2, z: 1 };
            assert.deepStrictEqual(helper.orderObject(unordered), expected);
        });

        it('should return an empty object if given an empty object', () => {
            const unordered = {};
            const expected = {};
            assert.deepStrictEqual(helper.orderObject(unordered), expected);
        });
    });

    //splitArrayIntoChunks()
    describe('#splitArrayIntoChunks', () => {
        it('should split an array into chunks of the given length', () => {
            const arr = [1, 2, 3, 4, 5, 6];
            const len = 2;
            const expected = [
                [1, 2],
                [3, 4],
                [5, 6],
            ];
            assert.deepStrictEqual(
                helper.splitArrayIntoChunks(arr, len),
                expected
            );
        });

        it('should return an array with a single chunk if the length is greater than the array length', () => {
            const arr = [1, 2, 3];
            const len = 5;
            const expected = [[1, 2, 3]];
            assert.deepStrictEqual(
                helper.splitArrayIntoChunks(arr, len),
                expected
            );
        });

        it('should return an empty array if given an empty array', () => {
            const arr = [];
            const len = 2;
            const expected = [];
            assert.deepStrictEqual(
                helper.splitArrayIntoChunks(arr, len),
                expected
            );
        });
    });

    //mergeArray()
    describe('#mergeArrays()', () => {
        it('should merge two arrays of objects with matching steamids', () => {
            const arr1 = [
                { steamid: '1', name: 'John' },
                { steamid: '2', name: 'Jane' },
                { steamid: '3', name: 'Bob' },
            ];
            const arr2 = [
                { steamid: '1', age: 30 },
                { steamid: '2', age: 25 },
            ];
            const expected = [
                { steamid: '1', name: 'John', age: 30 },
                { steamid: '2', name: 'Jane', age: 25 },
                { steamid: '3', name: 'Bob' },
            ];
            assert.deepStrictEqual(helper.mergeArrays(arr1, arr2), expected);
        });

        it('should return arr1 unchanged if arr2 is empty', () => {
            const arr1 = [
                { steamid: '1', name: 'John' },
                { steamid: '2', name: 'Jane' },
            ];
            const arr2 = [];
            assert.deepStrictEqual(helper.mergeArrays(arr1, arr2), arr1);
        });

        it('should return arr1 unchanged if no steamids match', () => {
            const arr1 = [
                { steamid: '1', name: 'John' },
                { steamid: '2', name: 'Jane' },
            ];
            const arr2 = [
                { steamid: '3', age: 30 },
                { steamid: '4', age: 25 },
            ];
            assert.deepStrictEqual(helper.mergeArrays(arr1, arr2), arr1);
        });
    });

    //getDiffOfArrs()
    describe('#getDiffOfArrs', () => {
        it('should return an array of elements that are in the first array but not the second', () => {
            const firstArr = [1, 2, 3, 4, 5];
            const secondArr = [3, 4, 5, 6];
            const expected = [1, 2];
            assert.deepStrictEqual(
                helper.getDiffOfArrs(firstArr, secondArr),
                expected
            );
        });

        it('should return an empty array if both arrays contain the same elements', () => {
            const firstArr = [1, 2, 3];
            const secondArr = [1, 2, 3];
            const expected = [];
            assert.deepStrictEqual(
                helper.getDiffOfArrs(firstArr, secondArr),
                expected
            );
        });

        it('should return the difference of arrays if they have a difference', () => {
            const firstArr = [1, 2, 3, 4, 5];
            const secondArr = [2, 3, 4];
            const expected = [1, 5];
            assert.deepStrictEqual(
                helper.getDiffOfArrs(firstArr, secondArr),
                expected
            );
        });

        it('should return the difference of arrays if they have a difference', () => {
            const firstArr = [2, 3, 4];
            const secondArr = [1, 2, 3, 4, 5];
            const expected = [1, 5];
            assert.deepStrictEqual(
                helper.getDiffOfArrs(firstArr, secondArr),
                expected
            );
        });

        it('should return the first array if the second array is empty', () => {
            const firstArr = [1, 2, 3];
            const secondArr = [];
            const expected = [1, 2, 3];
            assert.deepStrictEqual(
                helper.getDiffOfArrs(firstArr, secondArr),
                expected
            );
        });

        it('should return the second array if the first array is empty', () => {
            const firstArr = [];
            const secondArr = [1, 2, 3];
            const expected = [1, 2, 3];
            assert.deepStrictEqual(
                helper.getDiffOfArrs(firstArr, secondArr),
                expected
            );
        });
    });

    //downloadImage()
    describe('#downloadImage', () => {});
});


gulpfile.js
/* eslint-disable no-prototype-builtins */
/*!
 * ioBroker gulpfile
 * Date: 2019-01-28
 */
'use strict';

const gulp = require('gulp');
const fs = require('fs');
const pkg = require('./package.json');
const iopackage = require('./io-package.json');
const version = (pkg && pkg.version) ? pkg.version : iopackage.common.version;
const fileName = 'words.js';
const EMPTY = '';
const translate = require('./lib/tools').translateText;
const languages = {
    en: {},
    de: {},
    ru: {},
    pt: {},
    nl: {},
    fr: {},
    it: {},
    es: {},
    pl: {},
    uk: {},
    'zh-cn': {}
};

function lang2data(lang) {
    let str ='{\n';
    let count = 0;
    for (const w in lang) {
        if (lang.hasOwnProperty(w)) {
            count++;
            const key = '    "' + w.replace(/"/g, '\\"') + '": ';
            str += key + '"' + lang[w].replace(/"/g, '\\"') + '",\n';
        }
    }
    if (!count) {
        return '{\n}';
    } else {
        return str.substring(0, str.length - 2) + '\n}';
    }
}

function readWordJs(src) {
    try {
        let words;
        if (fs.existsSync(src + 'js/' + fileName)) {
            words = fs.readFileSync(src + 'js/' + fileName).toString();
        } else {
            words = fs.readFileSync(src + fileName).toString();
        }
        words = words.substring(words.indexOf('{'), words.length);
        words = words.substring(0, words.lastIndexOf(';'));

        const resultFunc = new Function('return ' + words + ';');

        return resultFunc();
    } catch (e) {
        return null;
    }
}

function padRight(text, totalLength) {
    return text + (text.length < totalLength ? new Array(totalLength - text.length).join(' ') : '');
}

function writeWordJs(data, src) {
    let text = '';
    text += '/*global systemDictionary:true */\n';
    text += "'use strict';\n\n";
    text += 'systemDictionary = {\n';
    for (const word in data) {
        if (data.hasOwnProperty(word)) {
            text += '    ' + padRight('"' + word.replace(/"/g, '\\"') + '": {', 50);
            let line = '';
            for (const lang in data[word]) {
                if (data[word].hasOwnProperty(lang)) {
                    line += '"' + lang + '": "' + padRight(data[word][lang].replace(/"/g, '\\"') + '",', 50) + ' ';
                }
            }
            if (line) {
                line = line.trim();
                line = line.substring(0, line.length - 1);
            }
            text += line + '},\n';
        }
    }
    text += '};';
    if (fs.existsSync(src + 'js/' + fileName)) {
        fs.writeFileSync(src + 'js/' + fileName, text);
    } else {
        fs.writeFileSync(src + '' + fileName, text);
    }
}

function words2languages(src) {
    const langs = Object.assign({}, languages);
    const data = readWordJs(src);
    if (data) {
        for (const word in data) {
            if (data.hasOwnProperty(word)) {
                for (const lang in data[word]) {
                    if (data[word].hasOwnProperty(lang)) {
                        langs[lang][word] = data[word][lang];
                        //  pre-fill all other languages
                        for (const j in langs) {
                            if (langs.hasOwnProperty(j)) {
                                langs[j][word] = langs[j][word] || EMPTY;
                            }
                        }
                    }
                }
            }
        }
        if (!fs.existsSync(src + 'i18n/')) {
            fs.mkdirSync(src + 'i18n/');
        }
        for (const l in langs) {
            if (!langs.hasOwnProperty(l))
                continue;
            const keys = Object.keys(langs[l]);
            keys.sort();
            const obj = {};
            for (let k = 0; k < keys.length; k++) {
                obj[keys[k]] = langs[l][keys[k]];
            }
            if (!fs.existsSync(src + 'i18n/' + l)) {
                fs.mkdirSync(src + 'i18n/' + l);
            }

            fs.writeFileSync(src + 'i18n/' + l + '/translations.json', lang2data(obj));
        }
    } else {
        console.error('Cannot read or parse ' + fileName);
    }
}

function languages2words(src) {
    const dirs = fs.readdirSync(src + 'i18n/');
    const langs = {};
    const bigOne = {};
    const order = Object.keys(languages);
    dirs.sort(function (a, b) {
        const posA = order.indexOf(a);
        const posB = order.indexOf(b);
        if (posA === -1 && posB === -1) {
            if (a > b)
                return 1;
            if (a < b)
                return -1;
            return 0;
        } else if (posA === -1) {
            return -1;
        } else if (posB === -1) {
            return 1;
        } else {
            if (posA > posB)
                return 1;
            if (posA < posB)
                return -1;
            return 0;
        }
    });
    for (const lang of dirs) {
        if (lang === 'flat.txt')
            continue;
        langs[lang] = fs.readFileSync(src + 'i18n/' + lang + '/translations.json').toString();
        langs[lang] = JSON.parse(langs[lang]);
        const words = langs[lang];
        for (const word in words) {
            if (words.hasOwnProperty(word)) {
                bigOne[word] = bigOne[word] || {};
                if (words[word] !== EMPTY) {
                    bigOne[word][lang] = words[word];
                }
            }
        }
    }
    // read actual words.js
    const aWords = readWordJs();

    const temporaryIgnore = ['flat.txt'];
    if (aWords) {
        // Merge words together
        for (const w in aWords) {
            if (aWords.hasOwnProperty(w)) {
                if (!bigOne[w]) {
                    console.warn('Take from actual words.js: ' + w);
                    bigOne[w] = aWords[w];
                }
                dirs.forEach(function (lang) {
                    if (temporaryIgnore.indexOf(lang) !== -1)
                        return;
                    if (!bigOne[w][lang]) {
                        console.warn('Missing "' + lang + '": ' + w);
                    }
                });
            }
        }

    }

    writeWordJs(bigOne, src);
}

async function translateNotExisting(obj, baseText, yandex) {
    let t = obj['en'];
    if (!t) {
        t = baseText;
    }

    if (t) {
        for (const l in languages) {
            if (!obj[l]) {
                const time = new Date().getTime();
                obj[l] = await translate(t, l, yandex);
                console.log('en -> ' + l + ' ' + (new Date().getTime() - time) + ' ms');
            }
        }
    }
}

//TASKS

gulp.task('adminWords2languages', function (done) {
    words2languages('./admin/');
    done();
});

gulp.task('adminLanguages2words', function (done) {
    languages2words('./admin/');
    done();
});

gulp.task('updatePackages', function (done) {
    iopackage.common.version = pkg.version;
    iopackage.common.news = iopackage.common.news || {};
    if (!iopackage.common.news[pkg.version]) {
        const news = iopackage.common.news;
        const newNews = {};

        newNews[pkg.version] = {
            en: 'news',
            de: 'neues',
            ru: 'новое',
            pt: 'novidades',
            nl: 'nieuws',
            fr: 'nouvelles',
            it: 'notizie',
            es: 'noticias',
            pl: 'nowości',
            uk: 'новини',
            'zh-cn': '新'
        };
        iopackage.common.news = Object.assign(newNews, news);
    }
    fs.writeFileSync('io-package.json', JSON.stringify(iopackage, null, 4));
    done();
});

gulp.task('updateReadme', function (done) {
    const readme = fs.readFileSync('README.md').toString();
    const pos = readme.indexOf('## Changelog\n');
    if (pos !== -1) {
        const readmeStart = readme.substring(0, pos + '## Changelog\n'.length);
        const readmeEnd = readme.substring(pos + '## Changelog\n'.length);

        if (readme.indexOf(version) === -1) {
            const timestamp = new Date();
            const date = timestamp.getFullYear() + '-' +
					('0' + (timestamp.getMonth() + 1).toString(10)).slice(-2) + '-' +
					('0' + (timestamp.getDate()).toString(10)).slice(-2);

            let news = '';
            if (iopackage.common.news && iopackage.common.news[pkg.version]) {
                news += '* ' + iopackage.common.news[pkg.version].en;
            }

            fs.writeFileSync('README.md', readmeStart + '### ' + version + ' (' + date + ')\n' + (news ? news + '\n\n' : '\n') + readmeEnd);
        }
    }
    done();
});

gulp.task('translate', async function () {

    let yandex;
    const i = process.argv.indexOf('--yandex');
    if (i > -1) {
        yandex = process.argv[i + 1];
    }

    if (iopackage && iopackage.common) {
        if (iopackage.common.news) {
            console.log('Translate News');
            for (const k in iopackage.common.news) {
                console.log('News: ' + k);
                const nw = iopackage.common.news[k];
                await translateNotExisting(nw, null, yandex);
            }
        }
        if (iopackage.common.titleLang) {
            console.log('Translate Title');
            await translateNotExisting(iopackage.common.titleLang, iopackage.common.title, yandex);
        }
        if (iopackage.common.desc) {
            console.log('Translate Description');
            await translateNotExisting(iopackage.common.desc, null, yandex);
        }

        if (fs.existsSync('./admin/i18n/en/translations.json')) {
            const enTranslations = require('./admin/i18n/en/translations.json');
            for (const l in languages) {
                console.log('Translate Text: ' + l);
                let existing = {};
                if (fs.existsSync('./admin/i18n/' + l + '/translations.json')) {
                    existing = require('./admin/i18n/' + l + '/translations.json');
                }
                for (const t in enTranslations) {
                    if (!existing[t]) {
                        existing[t] = await translate(enTranslations[t], l, yandex);
                    }
                }
                if (!fs.existsSync('./admin/i18n/' + l + '/')) {
                    fs.mkdirSync('./admin/i18n/' + l + '/');
                }
                fs.writeFileSync('./admin/i18n/' + l + '/translations.json', JSON.stringify(existing, null, 4));
            }
        }

    }
    fs.writeFileSync('io-package.json', JSON.stringify(iopackage, null, 4));
});

gulp.task('translateAndUpdateWordsJS', gulp.series('translate', 'adminLanguages2words', 'adminWords2languages'));

gulp.task('default', gulp.series('updatePackages', 'updateReadme'));

lib/helper.js
const axios = require('axios');

/**
 * Converts unix timestamp to '16 Aug 2018 20:39:09'
 * @function
 * @name timeConverter
 * @param {Number} timestamp
 * @returns {String}
 */
const timeConverter = (timestamp) => {
    if (isNaN(timestamp) || timestamp <= 0) {
        return '';
    }
    const time = new Date(timestamp * 1000);
    return time.toLocaleString();
};

/**
 * Sort keys of object
 * @function
 * @name orderObject
 * @param {Object} unordered
 * @returns {Object}
 */
const orderObject = (unordered) => {
    const ordered = Object.keys(unordered)
        .sort()
        .reduce((obj, key) => {
            obj[key] = unordered[key];
            return obj;
        }, {});
    // @ts-ignore
    return ordered;
};

/**
 * Split array into chunks
 * @function
 * @name splitArrayIntoChunks
 * @param {Array} arr
 * @param {Number} len
 * @returns {Array}
 */
const splitArrayIntoChunks = (arr, len) => {
    const chunks = [];
    const n = arr.length;
    let i = 0;
    while (i < n) chunks.push(arr.slice(i, (i += len)));
    return chunks;
};

/**
 * Merge 2 arrays with key
 * @function
 * @name mergeArrays
 * @param {Array} arr1
 * @param {Array} arr2
 * @returns {Array}
 */
const mergeArrays = (arr1, arr2) => {
    const merged = arr1.map((item) => {
        arr2.forEach((item2) => {
            if (item.steamid === item2.steamid) item = Object.assign(item, item2);
        });
        return item;
    });
    return merged;
};

/**
 * Returns difference between 2 arrays
 * @function
 * @name getDiffOfArrs
 * @param {Array} firstArr
 * @param {Array} secondArr
 * @returns {Array}
 */
const getDiffOfArrs = (firstArr, secondArr) => {
    const bigArr = firstArr.length >= secondArr.length ? firstArr : secondArr;
    const smallArr = firstArr.length < secondArr.length ? firstArr : secondArr;
    return bigArr.filter((x) => !smallArr.includes(x));
};

/**
 * Convert image from url to base64
 * @function
 * @name mergeArrays
 * @param {String} imageUrl
 * @returns {Promise<String>}
 */
const downloadImage = async (imageUrl) => {
    // @ts-ignore
    const response = await axios({
        method: 'GET',
        url: imageUrl,
        responseType: 'arraybuffer'
    });

    const base64Image = Buffer.from(response.data, 'binary').toString('base64');
    return `data:image/jpg;base64,${base64Image}`;
};

module.exports = { timeConverter, orderObject, splitArrayIntoChunks, mergeArrays, getDiffOfArrs, downloadImage };

main.js
'use strict';

/*
 * Created with @iobroker/create-adapter v1.34.1
 */

// The adapter-core module gives you access to the core ioBroker functions you need to create an adapter
const utils = require('@iobroker/adapter-core');
const helper = require('./lib/helper');
const axios = require('axios');

class SteamFriends extends utils.Adapter {
    /**
	 * @param {Partial<utils.AdapterOptions>} [options={}]
	 */
    constructor(options) {
        super({
            ...options,
            name: 'steam-friends',
        });
        this.on('ready', this.onReady.bind(this));
        this.on('unload', this.onUnload.bind(this));

        //Variables
        this.runInterval = null;
        this.isRunning = false;
    }
    /**
	 * Clean up the interval and outputs message if available
	 */
    cleanAllUp(message) {
        if (this.runInterval) {
            clearInterval(this.runInterval);
        }
        this.runInterval = null;
        if (message) {
            this.log.error(message);
        }
    }

    /**
	 * Get details of friends and return them as an array
	 */
    async getFriendDetails(connection, friends) {
        //Split the friends into chunks
        const friendsListChunks = helper.splitArrayIntoChunks(friends, 99);
        //Set needed variables
        const friendsDetail = [];
        const friendsWorking = [];

        //Get details for each chunk
        for (let i = 0; i < friendsListChunks.length; i++) {
            const chunk = friendsListChunks[i];
            const ids = chunk.map((ele) => ele.steamid).join(',');
            const url = `http://api.steampowered.com/ISteamUser/GetPlayerSummaries/v0002/?key=${connection.apikey}&steamids=${ids}`;
            try {
                // @ts-ignore
                const response = await axios.get(url);
                const data = response.data.response.players;
                const mergedData = helper.mergeArrays(chunk, data); //Merge 2 Arrays with same key (steamid)
                friendsWorking.push(...mergedData);
            } catch (error) {
                this.cleanAllUp(
                    `Could not load your friends data. Please try to restart the adapter!`
                );
                break;
            }
        }

        //Set object for each friend
        const states = {
            0: 'Offline',
            1: 'Online',
            2: 'Busy',
            3: 'Away',
            4: 'Snooze',
            5: 'looking to trade',
            6: 'looking to play',
        };
        for (let i = 0; i < friendsWorking.length; i++) {
            const ele = friendsWorking[i];
            const friend = {
                steamid: ele.steamid ? String(ele.steamid) : '',
                relationship: ele.relationship ? String(ele.relationship) : '',
                friend_since: ele.friend_since ? parseInt(ele.friend_since) : 0,
                friend_since_ts: ele.friend_since
                    ? helper.timeConverter(parseInt(ele.friend_since))
                    : '',
                personaname: ele.personaname ? String(ele.personaname) : '',
                profileurl: ele.profileurl ? String(ele.profileurl) : '',
                avatar: ele.avatar ? String(ele.avatar) : '',
                avatarmedium: ele.avatarmedium ? String(ele.avatarmedium) : '',
                avatarfull: ele.avatarfull ? String(ele.avatarfull) : '',
                avatarhash: ele.avatarhash ? String(ele.avatarhash) : '',
                avatarbase64: ele.avatarfull
                    ? await helper.downloadImage(String(ele.avatarfull))
                    : '',
                personastate: ele.personastate ? parseInt(ele.personastate) : 0,
                personastate_text: ele.personastate
                    ? states[parseInt(ele.personastate)]
                    : states[0],
                personastateflags: ele.personastateflags
                    ? parseInt(ele.personastateflags)
                    : 0,
                communityvisibilitystate: ele.communityvisibilitystate
                    ? parseInt(ele.communityvisibilitystate)
                    : 0,
                profilestate: ele.profilestate ? parseInt(ele.profilestate) : 0,
                lastlogoff: ele.lastlogoff ? parseInt(ele.lastlogoff) : 0,
                lastlogoff_ts: ele.lastlogoff
                    ? helper.timeConverter(parseInt(ele.lastlogoff))
                    : '',
                commentpermission: ele.commentpermission
                    ? parseInt(ele.commentpermission)
                    : 0,
                realname: ele.realname ? String(ele.realname) : '',
                primaryclanid: ele.primaryclanid
                    ? String(ele.primaryclanid)
                    : '',
                timecreated: ele.timecreated ? parseInt(ele.timecreated) : '',
                timecreated_ts: ele.timecreated
                    ? helper.timeConverter(parseInt(ele.timecreated))
                    : '',
                gameid: ele.gameid ? String(ele.gameid) : '',
                gameserverip: ele.gameserverip ? String(ele.gameserverip) : '',
                gameextrainfo: ele.gameextrainfo
                    ? String(ele.gameextrainfo)
                    : '',
                loccountrycode: ele.loccountrycode
                    ? String(ele.loccountrycode)
                    : '',
                locstatecode: ele.locstatecode ? String(ele.locstatecode) : '',
                loccityid: ele.loccityid ? String(ele.loccityid) : '',
            };
            friendsDetail.push(helper.orderObject(friend)); //Order keys of object
        }

        return friendsDetail;
    }

    async setFriendData(friend) {
        const channelId =
			friend.steamid === this.config.steamid ? 'me' : friend.steamid;
        const channelName = friend.personaname
            ? friend.personaname
            : friend.steamid;

        //Create the channel if it doesn't already exist
        this.setObjectNotExists(channelId, {
            type: 'channel',
            common: {
                name: channelName,
                icon: friend['avatarbase64'],
            },
            native: {},
        });

        //Set state for each key in object
        Object.getOwnPropertyNames(friend).forEach(async (key) => {
            const path = `${channelId}.${key}`;
            const state = await this.getStateAsync(path);

            //New state if it does not exists
            if (!state) {
                await this.setObjectNotExistsAsync(path, {
                    type: 'state',
                    common: {
                        name: key,
                        type: typeof friend[key],
                        role: 'state',
                        // @ts-ignore
                        read: true,
                        write: false,
                    },
                    native: {},
                });
            }

            //Set state value
            await this.setStateChangedAsync(path, friend[key], true);
        });
    }

    cleanUpFriends(friendlist) {
        // Only get steam-ids from the friends list
        const currentFriendlist = friendlist.map((ele) => {
            if (ele.steamid !== this.config.steamid && ele.steamid.length === 17) return String(ele.steamid);
        });

        // Get and remove all channels from the objects
        this.getChannelsOf((err, res) => {
            if (err || !res) {
                this.log.error('Could not delete former friends');
                return;
            }

            // Gets all channels with at least 16 numbers => all friends
            const channels = res.map((ele) => {
                const id = ele['_id'].split('.');
                const element = String(id[id.length - 1]);
                if (element && element.length === 17 && element !== this.config.steamid) return String(element);
            });

            // Gets difference between the channels and the current friends list
            const difference = helper.getDiffOfArrs(
                channels,
                currentFriendlist
            );
            // Remove old "friends" / channels
            for (let index = 0; index < difference.length; index++) {
                const id = difference[index];
                this.delObject(id, { recursive: true });
            }
        });
    }

    /**
	 * Is called when databases are connected and adapter received configuration.
	 * Initializing the adapter
	 */
    async onReady() {
        this.setState('info.connection', false, true); // Reset the connection indicator during startup

        //Checks at the beginning
        if (this.config.refreshinterval < 10) {
            this.cleanAllUp(
                `Refresh interval is smaller than 10 seconds (Current: ${this.config.refreshinterval})`
            );
            return;
        }

        //Set variables from config
        const refreshInterval = this.config.refreshinterval * 1000;
        const connection = {
            steamid: this.config.steamid,
            apikey: this.config.apikey,
        };

        // Main program
        this.runInterval = setInterval(async () => {
            //Get friends list
            let friends = null;
            try {
                // @ts-ignore
                const response = await axios.get(
                    `http://api.steampowered.com/ISteamUser/GetFriendList/v0001/?key=${connection.apikey}&steamid=${connection.steamid}&relationship=friend`
                );
                friends = response.data.friendslist.friends;
                //Check if you have some friends ;)
                if (friends.length === 0) {
                    this.cleanAllUp(
                        `No friends found. Please verify your Steam-ID and check if your friends list is public.`
                    );
                    return;
                }
            } catch (error) {
                //Error if service is unavailable or the API key is wrong
                this.cleanAllUp(
                    `Access is denied. Retrying will not help. Please verify your API-Key.`
                );
                return;
            }

            //Add the own steamid to the list and get friend details
            friends.push({
                steamid: connection.steamid,
            });
            const friendsDetail = await this.getFriendDetails(
                connection,
                friends
            );

            //Now we have all data and need to add it as ioBroker states
            for (let i = 0; i < friendsDetail.length; i++) {
                const friend = friendsDetail[i];
                await this.setFriendData(friend);
            }

            //Deletes the channels of friends you don't have anymore
            this.cleanUpFriends(friends);
        }, refreshInterval);

        this.setState('info.connection', true, true);
    }

    /**
	 * Is called when adapter shuts down - callback has to be called under any circumstances!
	 * @param {() => void} callback
	 */
    onUnload(callback) {
        try {
            this.cleanAllUp();

            callback();
        } catch (e) {
            callback();
        }
    }
}

if (require.main !== module) {
    // Export the constructor in compact mode
    /**
	 * @param {Partial<utils.AdapterOptions>} [options={}]
	 */
    module.exports = (options) => new SteamFriends(options);
} else {
    // otherwise start the instance directly
    new SteamFriends();
}
